packages/interbit-covenant-tools/src/rootCovenant/covenants/manifestCovenant/applyJoinChanges.js

```diff
@@ -0,0 +1,170 @@
+const _ = require('lodash')
+const {
+  startConsumeState,
+  startProvideState,
+  authorizeReceiveActions,
+  authorizeSendActions
+} = require('interbit-covenant-utils')
+const {
+  redispatch,
+  actionTypes: coreActionTypes
+} = require('../../../coreCovenant')
+const {
+  stopConsumeState,
+  stopProvideState,
+  revokeSendActions,
+  revokeReceiveActions
+} = require('../../../coreCovenant/helpers/revokeJoin')
+const { JOIN_TYPES } = require('../../../constants')
+
+const applyJoinChanges = (state, newManifest) => {
+  let nextState = state
+
+  const newJoinActions = manifestJoinsToActions(newManifest)
+  if (!state.manifest) {
+    for (const joinAction of newJoinActions) {
+      nextState = redispatch(nextState, joinAction)
+    }
+
+    return nextState
+  }
+
+  const currManifest = !state.manifest.chainId
+    ? Object.values(state.manifest.manifest)[0]
+    : state.manifest
+
+  const currJoinActions = manifestJoinsToActions(currManifest)
+
+  const removedJoinActions = currJoinActions.filter(
+    currJoin => !newJoinActions.some(newJoin => _.isEqual(currJoin, newJoin))
+  )
+  for (const joinAction of removedJoinActions) {
+    switch (joinAction.type) {
+      case coreActionTypes.AUTHORIZE_RECEIVE_ACTIONS: {
+        const { permittedActions, senderChainId } = joinAction.payload
+        for (const permittedAction of permittedActions) {
+          nextState = revokeReceiveActions(
+            nextState,
+            senderChainId,
+            permittedAction
+          )
+        }
+        break
+      }
+
+      case coreActionTypes.AUTHORIZE_SEND_ACTIONS: {
+        const { receiverChainId } = joinAction.payload
+        nextState = revokeSendActions(nextState, receiverChainId)
+        break
+      }
+
+      case coreActionTypes.START_PROVIDE_STATE: {
+        const { joinName } = joinAction.payload
+        nextState = stopProvideState(nextState, joinName)
+        break
+      }
+
+      case coreActionTypes.START_CONSUME_STATE: {
+        const { joinName } = joinAction.payload
+        nextState = stopConsumeState(nextState, joinName)
+        break
+      }
+
+      default:
+        break
+    }
+  }
+
+  const addedJoinActions = newJoinActions.filter(
+    newJoin => !currJoinActions.some(currJoin => _.isEqual(newJoin, currJoin))
+  )
```

Comment:
This filter condition is virtually the same as lines 38-40. This should be factored out.